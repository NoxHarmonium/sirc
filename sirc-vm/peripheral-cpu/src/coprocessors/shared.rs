use peripheral_bus::{device::BusAssertions, BusPeripheral};

use crate::registers::{ExceptionUnitRegisters, Registers, SegmentedAddress};

#[derive(Default, FromPrimitive, ToPrimitive, Debug, PartialEq, Eq)]
pub enum ExecutionPhase {
    #[default]
    InstructionFetchLow = 0x0,
    InstructionFetchHigh = 0x1,
    InstructionDecode = 0x2,
    ExecutionEffectiveAddressExecutor = 0x3,
    MemoryAccessExecutor = 0x4,
    WriteBackExecutor = 0x5,
}

pub trait Executor {
    const COPROCESSOR_ID: u8;

    fn step<'a>(
        &mut self,
        phase: &ExecutionPhase,
        cause_register_value: u16,
        registers: &'a mut Registers,
        eu_registers: &'a mut ExceptionUnitRegisters,
        bus_assertions: BusAssertions,
    ) -> BusAssertions;
}

/// Extends an 8 bit value that is signed to a 16 bit value
///
/// Most of the values we pass around in the virtual machine are represented as unsigned
/// (even if they are technically 2's compliment signed)
/// This is fine most of the time, however, if we want to cast a smaller value to a
/// larger value, if the compiler doesn't know that the bits actually represent a
/// 2's complement number, it will fill the new space with zeros which will cause
/// the number's value to change.
///
///
/// ```
/// use peripheral_cpu::coprocessors::shared::sign_extend_small_offset;
///
/// assert_eq!(sign_extend_small_offset(0xFF), 0xFFFF);
/// assert_eq!(sign_extend_small_offset(0x00), 0x0000);
/// ```
///
#[must_use]
#[allow(clippy::cast_sign_loss, clippy::cast_possible_wrap)]
pub fn sign_extend_small_offset(small_offset: u8) -> u16 {
    // 1. Convert to i8 so the compiler knows we are dealing with a signed number
    // 2. Extend it to i16, the compiler knows it is signed so it preserves the sign
    // 3. Cast it back to unsigned which we deal with
    i16::from(small_offset as i8) as u16
}

// TODO: Rename to something more generic like fetch double word
#[must_use]
pub fn fetch_instruction(mem: &mut BusPeripheral, pc: (u16, u16)) -> [u8; 4] {
    // Only the CPU knows that the address is split into two 16 bit registers
    // Any other peripheral will only see the 24 address lines
    let full_address: u32 = pc.to_full_address();

    // TODO: Alignment check?
    // TODO: Do we need to copy here?
    let [b1, b2] = u16::to_be_bytes(mem.read_address(full_address).to_owned());
    let [b3, b4] = u16::to_be_bytes(mem.read_address(full_address + 1).to_owned());
    [b1, b2, b3, b4]
}
