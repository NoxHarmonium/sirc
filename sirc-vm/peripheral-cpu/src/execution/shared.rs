use peripheral_mem::MemoryPeripheral;

use crate::{
    instructions::definitions::{ConditionFlags, Instruction, StatusRegisterUpdateSource},
    registers::Registers,
};

/**
* The instruction mapped out into components.
*
* Simulates the temporary registers the CPU would have when an instruction
* is being decoded.
*
* To avoid microcode/branching etc. all instructions are mapped out to the the
* same set of registers, however, depending on the instruction, some of the
* fields might be zero or full of garbage. You will need to make sure
* you know what instruction you are using before interpreting these
* registers.
*
* Future work: it might be a good idea in the future to type this so
* only the relevant registers are available for each instruction type.

*/
#[derive(Debug, Default, PartialEq, Eq)]
pub struct DecodedInstruction {
    // Raw Instruction Decode
    pub ins: Instruction,
    pub des: u8,
    pub sr_a: u8,
    pub sr_b: u8,
    pub con: ConditionFlags,
    pub adr: u8,
    pub sr_src: StatusRegisterUpdateSource,
    // Inferred
    pub ad_l: u8,
    pub ad_h: u8,
    pub addr_inc: i8,
    pub des_ad_l: u8,
    pub des_ad_h: u8,
    pub sr_shift: u16,
    // Dereferenced
    pub sr_a_: u16,
    pub sr_b_: u16,
    pub ad_l_: u16,
    pub ad_h_: u16,
    pub con_: bool,
    pub sr: u16,
    pub npc_l_: u16,
    pub npc_h_: u16,
}

#[derive(Debug, Default)]
pub struct IntermediateRegisters {
    pub alu_output: u16,
    pub alu_status_register: u16,
    pub lmd: u16,
}

pub trait StageExecutor {
    fn execute(
        decoded_instruction: &DecodedInstruction,
        registers: &mut Registers,
        intermediate_registers: &mut IntermediateRegisters,
        // TODO: Only the memory access stage needs this. Maybe there is a clever way to only provide each stage what they need?
        mem: &MemoryPeripheral,
    );
}

pub enum ExecutionStage {
    // TODO: Work out a cleaner way to specify the data in each stage
    Execution,
    MemoryAccessAndBranchCompletion,
    WriteBack,
}
