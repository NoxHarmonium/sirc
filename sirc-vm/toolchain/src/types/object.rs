use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, Copy, Eq, PartialEq)]
pub enum RefType {
    /// A 16-bit signed offset from the current PC
    Offset,
    /// An 8 bit signed offset from the current PC (used for short immediate instructions)
    SmallOffset,
    /// The lower 16 bits of a full 24-bit address of the target
    LowerWord,
    /// The upper 8 bits of a full 24-bit address of the target (8 bits are ignored)
    UpperWord,
    /// Full 24 bit program address. Only supported with the DQ directive, can not fit in instructions
    FullAddress,
    /// Automatically determine the ref type based on the instruction
    Implied,
}

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct SymbolDefinition {
    pub name: String,
    pub offset: u32,
}

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct SymbolRef {
    pub name: String,
    pub offset: u32,
    pub ref_type: RefType,
    pub data_only: bool,
}

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct ObjectDefinition {
    // The offset in these definitions is the location of the symbol
    pub symbols: Vec<SymbolDefinition>,
    // The offset in these definitions is the location of the ref in the program
    pub symbol_refs: Vec<SymbolRef>,
    pub program: Vec<u8>,
}

pub fn merge_object_definitions(object_definitions: &[ObjectDefinition]) -> ObjectDefinition {
    let empty_definition: ObjectDefinition = ObjectDefinition {
        symbols: vec![],
        symbol_refs: vec![],
        program: vec![],
    };

    object_definitions
        .iter()
        .fold(empty_definition, |prev, curr| -> ObjectDefinition {
            let prev_offset: u32 = prev
                .program
                .len()
                .try_into()
                .expect("Program length cannot not be larger than 32 bits");

            let offset_symbols: Vec<SymbolDefinition> = curr
                .symbols
                .iter()
                .map(|s| SymbolDefinition {
                    name: s.name.clone(),
                    offset: prev_offset + s.offset,
                })
                .collect();

            let offset_symbol_refs: Vec<SymbolRef> = curr
                .symbol_refs
                .iter()
                .map(|s| SymbolRef {
                    name: s.name.clone(),
                    offset: prev_offset + s.offset,
                    ..*s
                })
                .collect();

            ObjectDefinition {
                symbols: [prev.symbols.as_slice(), offset_symbols.as_slice()].concat(),
                symbol_refs: [prev.symbol_refs.as_slice(), offset_symbol_refs.as_slice()].concat(),
                program: [prev.program.as_slice(), curr.program.as_slice()].concat(),
            }
        })
}
