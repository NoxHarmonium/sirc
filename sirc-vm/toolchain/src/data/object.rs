use peripheral_cpu::coprocessors::processing_unit::definitions::{
    ImmediateInstructionData, InstructionData, ShortImmediateInstructionData,
    INSTRUCTION_SIZE_BYTES,
};

use crate::parsers::data::DataType;
use crate::parsers::instruction::{DataToken, Token};
use crate::types::object::{ObjectDefinition, SymbolDefinition, SymbolRef};
use peripheral_cpu::coprocessors::processing_unit::encoding::encode_instruction;

use std::collections::HashMap;

fn resolve_placeholder(
    placeholders: &HashMap<String, u32>,
    placeholder_name: &String,
    instruction_data: &InstructionData,
) -> InstructionData {
    let resolved_value = placeholders.get(placeholder_name)
        .unwrap_or_else(|| panic!("Could not find a value for placeholder name [{placeholder_name}]. Make sure it is defined with the .EQU directive."));
    match instruction_data {
        InstructionData::Immediate(immediate_instruction) => {
            if resolved_value > &0xFFFF {
                panic!("Immediate value (resolved from [{placeholder_name}] placeholder) can only be up to 16 bits ({resolved_value} > 0xFFFF)");
            } else {
                InstructionData::Immediate(ImmediateInstructionData {
                    value: resolved_value
                        .to_owned()
                        .try_into()
                        .expect("Value should fit into a u8 as it is checked above"),
                    ..immediate_instruction.clone()
                })
            }
        }
        InstructionData::ShortImmediate(short_immediate_instruction) => {
            if resolved_value > &0xFF {
                panic!("Immediate value (resolved from [{placeholder_name}] placeholder) can only be up to 8 bits when using a shift definition ({resolved_value} > 0xFF)");
            } else {
                InstructionData::ShortImmediate(ShortImmediateInstructionData {
                    value: resolved_value
                        .to_owned()
                        .try_into()
                        .expect("Value should fit into a u8 as it is checked above"),
                    ..short_immediate_instruction.clone()
                })
            }
        }
        InstructionData::Register(_) => instruction_data.clone(),
    }
}

#[allow(clippy::cast_possible_truncation)]
fn inject_data_value(
    data: DataToken,
    program: &mut [[u8; 4]],
    program_offset: usize,
    symbol_refs: &mut Vec<SymbolRef>,
    offset: u32,
    placeholders: &HashMap<String, u32>,
) {
    match data.value {
        DataType::Value(value) => {
            // Align on instruction sizes
            // TODO: Make packing smaller data sizes more efficient
            // E.g. put 4 DBs in one 32 bit chunk
            // TODO: Clean up this mess
            let bytes: [u8; 4] = match data.size_bytes {
                1 => [0x0, 0x0, 0x0, value as u8],
                2 => {
                    let word_bytes = u16::to_be_bytes(value as u16);
                    [0x0, 0x0, word_bytes[0], word_bytes[1]]
                }
                4 => u32::to_be_bytes(value),
                _ => panic!("Unsupported data size bytes {}", data.size_bytes),
            };

            program[program_offset] = bytes;
        }
        DataType::SymbolRef(symbol_ref) => {
            program[program_offset] = [0x0, 0x0, 0x0, 0x0];
            symbol_refs.push(SymbolRef {
                name: symbol_ref.name,
                offset,
                ref_type: symbol_ref.ref_type,
                data_only: true,
            });
        }
        DataType::PlaceHolder(placeholder_name) => {
            let resolved_value = placeholders.get(&placeholder_name)
                .unwrap_or_else(|| panic!("Could not find a value for placeholder name [{placeholder_name}]. Make sure it is defined with the .EQU directive."));
            program[program_offset] = u32::to_be_bytes(resolved_value.to_owned());
        }
    }
}

fn ensure_program_size(program: &mut Vec<[u8; 4]>, min_size: usize) {
    if min_size > program.len() {
        program.resize(min_size + 1, [0x0, 0x0, 0x0, 0x0]);
    }
}

pub fn build_object(tokens: Vec<Token>) -> ObjectDefinition {
    let mut symbols: Vec<SymbolDefinition> = vec![];
    let mut symbol_refs: Vec<SymbolRef> = vec![];
    let mut placeholders: HashMap<String, u32> = HashMap::new();
    let mut offset: u32 = 0x0;
    let mut program: Vec<[u8; 4]> = vec![];

    for token in tokens {
        let program_offset: usize = offset as usize / 4;
        ensure_program_size(&mut program, program_offset + 1);

        match token {
            Token::Instruction(data) => {
                if let Some(symbol_ref) = data.symbol_ref {
                    symbol_refs.push(SymbolRef {
                        name: symbol_ref.name,
                        offset,
                        ref_type: symbol_ref.ref_type,
                        data_only: false,
                    });
                }

                let instruction = if let Some(placeholder_name) = data.placeholder_name {
                    resolve_placeholder(&placeholders, &placeholder_name, &data.instruction)
                } else {
                    data.instruction
                };

                program[program_offset] = encode_instruction(&instruction);

                offset += INSTRUCTION_SIZE_BYTES;
            }
            Token::Label(data) => symbols.push(SymbolDefinition {
                name: data.name,
                offset,
            }),
            Token::Comment => {
                // Do nothing.
            }
            Token::Origin(data) => {
                // Word based addressing to match CPU
                offset = data.offset * 2;
            }
            Token::Data(data) => {
                inject_data_value(
                    data,
                    &mut program,
                    program_offset,
                    &mut symbol_refs,
                    offset,
                    &placeholders,
                );
                offset += INSTRUCTION_SIZE_BYTES;
            }
            Token::Equ(data) => {
                placeholders.insert(data.placeholder_name, data.value);
            }
        }
    }

    let bytes: Vec<u8> = program
        .iter()
        .flat_map(std::borrow::ToOwned::to_owned)
        .collect();

    ObjectDefinition {
        symbols,
        symbol_refs,
        program: bytes,
    }
}
