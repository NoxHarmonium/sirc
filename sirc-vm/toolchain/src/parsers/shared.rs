use nom::branch::alt;
use nom::bytes::complete::is_a;
use nom::character::complete::{char, digit1, one_of, space0};
use nom::combinator::{cut, map, map_res, opt, recognize};
use nom::error::{ErrorKind, ParseError};
use nom::sequence::{pair, preceded, terminated, tuple};
use nom::{AsChar, IResult};
use nom::{Err, InputTakeAtPosition};
use nom_supreme::error::ErrorTree;
use nom_supreme::error::{BaseErrorKind, Expectation};
use nom_supreme::tag::complete::tag;
use peripheral_cpu::coprocessors::processing_unit::definitions::{ShiftOperand, ShiftType};

use crate::types::object::RefType;

use super::instruction::{RefToken, ShiftDefinitionData};

pub type AsmResult<'a, 'b, O> = IResult<&'a str, O, ErrorTree<&'b str>>;

#[allow(clippy::type_repetition_in_bounds)]
pub fn lexeme<'a, F: 'a, O, E: ParseError<&'a str>>(
    inner: F,
) -> impl FnMut(&'a str) -> IResult<&'a str, O, E>
where
    F: FnMut(&'a str) -> IResult<&'a str, O, E>,
{
    terminated(inner, space0)
}

fn parse_label_name_(i: &str) -> AsmResult<&str> {
    i.split_at_position1_complete(
        |item| !item.is_alphanum() && item != '_',
        ErrorKind::AlphaNumeric,
    )
}

fn parse_hex_<T: num_traits::Num>(i: &str) -> AsmResult<T> {
    let (i, _) = tag("0x")(i)?;
    let (i, raw_digits) = is_a(&b"0123456789abcdefABCDEF"[..])(i)?;
    let hex_parse_result = T::from_str_radix(raw_digits, 16);
    hex_parse_result.map_or_else(
        |_| {
            Err(Err::Error(ErrorTree::Base {
                location: i,
                kind: BaseErrorKind::Expected(Expectation::HexDigit),
            }))
        },
        |hex_value| Ok((i, hex_value)),
    )
}

// TODO: Allow u32 decimal
#[allow(clippy::cast_sign_loss)]
fn parse_dec_(i: &str) -> AsmResult<u32> {
    map_res(
        tuple((opt(one_of("+-")), recognize(digit1))),
        |(sign, number_string)| {
            sign.map_or_else(
                || str::parse::<u32>(number_string),
                |sign_value| {
                    // TODO: Re-concatenating the original string seems bad
                    // We should probably just get the original value or something
                    let full_number = format!("{sign_value}{number_string}");
                    // Signed numbers represented in parser as unsigned for simplicity
                    str::parse::<i32>(full_number.as_str()).map(|signed| signed as u32)
                },
            )
        },
    )(i)
}

fn parse_number_(i: &str) -> AsmResult<u32> {
    preceded(char('#'), alt((parse_hex, parse_dec)))(i)
}

fn parse_comma_sep_(i: &str) -> AsmResult<()> {
    let (i, (_, _)) = pair(char(','), space0)(i)?;
    Ok((i, ()))
}

pub fn parse_label_(i: &str) -> AsmResult<&str> {
    preceded(char(':'), cut(parse_label_name_))(i)
}

pub fn parse_origin_(i: &str) -> AsmResult<u32> {
    let (i, (_, value)) = tuple((lexeme(tag(".ORG")), alt((parse_hex, parse_dec))))(i)?;

    Ok((i, value))
}

pub fn parse_symbol_reference_postamble_(i: &str) -> AsmResult<Option<RefType>> {
    let (i, dot_parsed) = opt(tag("."))(i)?;

    match dot_parsed {
        Some(_) => map(
            alt((tag(".r"), tag(".u"), tag(".l"))),
            |parsed_value| match parsed_value {
                ".r" => Some(RefType::Offset),
                ".u" => Some(RefType::UpperWord),
                ".l" => Some(RefType::LowerWord),
                // TODO: Proper error handling again
                _ => panic!("Unknown postamble {parsed_value}"),
            },
        )(i),
        None => Ok((i, None)),
    }
}

pub fn parse_symbol_reference_(i: &str) -> AsmResult<RefToken> {
    let (i, name) = preceded(char('@'), parse_label_name_)(i)?;

    let (i, optional_preamble) = parse_symbol_reference_postamble_(i)?;

    Ok((
        i,
        RefToken {
            name: String::from(name),
            ref_type: optional_preamble.unwrap_or(RefType::Implied),
        },
    ))
}

pub fn parse_hex(i: &str) -> AsmResult<u32> {
    lexeme(parse_hex_)(i)
}

pub fn parse_dec(i: &str) -> AsmResult<u32> {
    lexeme(parse_dec_)(i)
}

pub fn parse_number(i: &str) -> AsmResult<u32> {
    lexeme(parse_number_)(i)
}

pub fn parse_comma_sep(i: &str) -> AsmResult<()> {
    lexeme(parse_comma_sep_)(i)
}

pub fn parse_label(i: &str) -> AsmResult<&str> {
    lexeme(parse_label_)(i)
}

pub fn parse_origin(i: &str) -> AsmResult<u32> {
    lexeme(parse_origin_)(i)
}

pub fn parse_symbol_reference(i: &str) -> AsmResult<RefToken> {
    lexeme(parse_symbol_reference_)(i)
}

///
/// Takes the parser representation of a shift and splits it out into the components
/// that get encoded into instructions.
///
/// In this case of instructions, the "`shift_count`" is either the index of the register,
/// or a constant depending on the value of `ShiftOperand`.
///
/// TODO: Should shift be stored as an enum in the instruction structs? then it could be reused
/// by the parser and avoid this function
///
pub fn split_shift_definition_data(
    shift_definition_data: &ShiftDefinitionData,
) -> (ShiftOperand, ShiftType, u8) {
    match shift_definition_data {
        // TODO: Probably can avoid this wrapping/unwrapping by using one type or something?
        crate::parsers::instruction::ShiftDefinitionData::Immediate(shift_type, shift_count) => {
            (ShiftOperand::Immediate, *shift_type, *shift_count)
        }
        crate::parsers::instruction::ShiftDefinitionData::Register(shift_type, shift_count) => (
            ShiftOperand::Register,
            *shift_type,
            shift_count.to_register_index(),
        ),
    }
}
