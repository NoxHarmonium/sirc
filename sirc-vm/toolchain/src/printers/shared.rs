use crate::printers::data::{print_data_token, print_equ_token};
use crate::types::data::RefToken;
use crate::types::object::RefType;
use crate::types::shared::{
    LabelToken, NumberToken, NumberType, Token, REF_TOKEN_LOWER_WORD_SUFFIX,
    REF_TOKEN_OFFSET_SUFFIX, REF_TOKEN_UPPER_WORD_SUFFIX,
};
use itertools::Itertools;

/// Prints the AST representation of a `NumberToken` to a string
///
///```
/// use toolchain::printers::shared::print_number_token;
/// use toolchain::types::shared::{NumberToken, NumberType};
/// let printed = print_number_token(&NumberToken {
///    value: 0xCAFE,
///    number_type: NumberType::Decimal
/// });
/// assert_eq!(String::from("#51966"), printed);
/// ```
pub fn print_number_token(number_token: &NumberToken) -> String {
    match number_token.number_type {
        NumberType::Hex => {
            if number_token.value > 0xFFFF {
                format!("#0x{:08X}", number_token.value)
            } else {
                format!("#0x{:04X}", number_token.value)
            }
        }
        NumberType::Decimal => format!("#{}", number_token.value),
    }
}

/// Prints the AST representation of a `RefToken` to a string
///
///```
/// use toolchain::printers::data::print_equ_token;
/// use toolchain::printers::shared::print_ref_token;
/// use toolchain::types::data::{EquToken, RefToken};
/// use toolchain::types::object::RefType;
/// use toolchain::types::shared::{NumberToken, NumberType};
/// let printed = print_ref_token(&RefToken {
///    name: String::from("some_other_symbol"),
///    ref_type: RefType::UpperWord
/// });
/// assert_eq!(String::from("@some_other_symbol.u"), printed);
/// ```
pub fn print_ref_token(ref_token: &RefToken) -> String {
    let optional_postamble = match ref_token.ref_type {
        RefType::Offset => String::from(REF_TOKEN_OFFSET_SUFFIX),
        RefType::LowerWord => String::from(REF_TOKEN_LOWER_WORD_SUFFIX),
        RefType::UpperWord => String::from(REF_TOKEN_UPPER_WORD_SUFFIX),
        _ => String::from(""),
    };

    return format!("@{}{}", ref_token.name, optional_postamble);
}

/// Prints an AST `Token` to a string
///
/// This is the generic function that should work over any token
///
///```
/// use toolchain::printers::shared::print_token;
/// use toolchain::types::data::{EquToken, RefToken};
/// use toolchain::types::object::RefType;
/// use toolchain::types::shared::{NumberToken, NumberType, Token};
/// let printed = print_token(&Token::Equ(
///    EquToken {
///        placeholder_name: String::from("BAR"),
///        number_token: NumberToken {
///            number_type: NumberType::Decimal,
///            value: 1234
///        }
///    }
/// ));
/// assert_eq!(String::from(".EQU $BAR #1234"), printed);
/// ```
pub fn print_token(token: &Token) -> String {
    match token {
        Token::Comment(text) => format!(";{text}"),
        Token::Label(LabelToken { name }) => format!(":{name}"),
        Token::Instruction(_) => todo!(),
        Token::Origin(_) => todo!(),
        Token::Data(data_token) => print_data_token(data_token),
        Token::Equ(equ_token) => print_equ_token(equ_token),
    }
}

/// Prints an AST tree to a string
///
///```
/// use toolchain::printers::shared::print_tokens;
/// use toolchain::types::data::{EquToken, RefToken};
/// use toolchain::types::object::RefType;
/// use toolchain::types::shared::{NumberToken, NumberType, Token};
/// let printed = print_tokens(vec![
///    Token::Equ(
///        EquToken {
///            placeholder_name: String::from("FOO"),
///            number_token: NumberToken {
///                number_type: NumberType::Decimal,
///                value: 1234
///            }
///        }
///    ),
///    Token::Equ(
///        EquToken {
///            placeholder_name: String::from("BAR"),
///            number_token: NumberToken {
///                number_type: NumberType::Hex,
///                value: 0xCAFE
///            }
///        }
///    ),
/// ]);
/// assert_eq!(r".EQU $FOO #1234
///.EQU $BAR #0xCAFE
///", printed);
/// ```
pub fn print_tokens(tokens: Vec<Token>) -> String {
    // TODO: Don't discard whitespace when printing
    // category=Toolchain
    // If you parse a file and then print it out again, you lose all the vertical whitespace
    tokens.iter().map(|token| print_token(token)).join("\n") + "\n"
}
