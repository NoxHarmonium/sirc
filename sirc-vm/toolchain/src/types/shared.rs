use crate::types::data::{DataToken, EquToken};
use crate::types::instruction::InstructionToken;
use serde::Serialize;

pub const REF_TOKEN_OFFSET_SUFFIX: &'static str = ".r";
pub const REF_TOKEN_LOWER_WORD_SUFFIX: &'static str = ".l";
pub const REF_TOKEN_UPPER_WORD_SUFFIX: &'static str = ".u";

#[derive(Debug, Serialize)]
pub enum NumberType {
    Hex,
    Decimal,
}

#[derive(Debug, Serialize)]
pub struct NumberToken {
    pub value: u32,
    pub number_type: NumberType,
}

#[derive(Debug, Serialize)]
pub struct LabelToken {
    pub name: String,
}

#[derive(Debug, Serialize)]
pub struct OriginToken {
    pub offset: u32,
}

#[derive(Debug)]
pub enum Token {
    Comment(String),
    Label(LabelToken),
    Instruction(InstructionToken),
    Origin(OriginToken),
    Data(DataToken),
    Equ(EquToken),
}
