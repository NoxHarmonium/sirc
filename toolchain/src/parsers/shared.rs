use nom::branch::alt;
use nom::bytes::complete::{is_a, tag};
use nom::character::complete::{alphanumeric1, digit1, multispace0, one_of, space0};
use nom::combinator::{map, map_res, opt, recognize};
use nom::error::{ErrorKind, ParseError};
use nom::sequence::{pair, terminated, tuple};
use nom::Err;
use nom::IResult;

use crate::types::object::RefType;

use super::instruction::RefToken;

pub fn lexeme<'a, F: 'a, O, E: ParseError<&'a str>>(
    inner: F,
) -> impl FnMut(&'a str) -> IResult<&'a str, O, E>
where
    F: FnMut(&'a str) -> IResult<&'a str, O, E>,
{
    terminated(inner, multispace0)
}

fn parse_hex_(i: &str) -> IResult<&str, u16> {
    let (i, _) = tag("0x")(i)?;
    let (i, raw_digits) = is_a(&b"0123456789abcdefABCDEF"[..])(i)?;
    let hex_parse_result = u16::from_str_radix(raw_digits, 16);
    match hex_parse_result {
        Ok(hex_value) => Ok((i, hex_value)),
        Err(_) => Err(Err::Error(nom::error::Error {
            input: i,
            code: ErrorKind::Fail,
        })),
    }
}

fn parse_dec_(i: &str) -> IResult<&str, u16> {
    map_res(
        tuple((opt(one_of("+-")), recognize(digit1))),
        |(sign, number_string)| {
            if sign.is_some() {
                // Signed numbers represented in parser as unsigned for simplicity
                str::parse::<i16>(number_string).map(|signed| signed as u16)
            } else {
                str::parse::<u16>(number_string)
            }
        },
    )(i)
}

fn parse_number_(i: &str) -> IResult<&str, u16> {
    alt((parse_hex, parse_dec))(i)
}

fn parse_comma_sep_(i: &str) -> IResult<&str, ()> {
    let (i, (_, _)) = pair(tag(","), space0)(i)?;
    Ok((i, ()))
}

fn parse_range_sep_(i: &str) -> IResult<&str, ()> {
    let (i, (_, _)) = pair(tag("->"), space0)(i)?;
    Ok((i, ()))
}

pub fn parse_label_(i: &str) -> IResult<&str, &str> {
    let (i, _) = tag(":")(i)?;
    alphanumeric1(i)
}

pub fn parse_symbol_reference_postamble_(i: &str) -> IResult<&str, Option<RefType>> {
    let (i, dot_parsed) = opt(tag("."))(i)?;

    match dot_parsed {
        Some(_) => map(
            alt((tag(".r"), tag(".u"), tag(".l"))),
            |parsed_value| match parsed_value {
                ".r" => Some(RefType::Offset),
                ".u" => Some(RefType::UpperByte),
                ".l" => Some(RefType::LowerByte),
                // TODO: Proper error handling again
                _ => panic!("Unknown postamble {}", parsed_value),
            },
        )(i),
        None => Ok((i, None)),
    }
}

pub fn parse_symbol_reference_(i: &str) -> IResult<&str, RefToken> {
    let (i, _) = tag("@")(i)?;
    let (i, name) = alphanumeric1(i)?;
    let (i, optional_preamble) = parse_symbol_reference_postamble_(i)?;

    Ok((
        i,
        RefToken {
            name: String::from(name),
            ref_type: optional_preamble.unwrap_or(RefType::Implied),
        },
    ))
}

pub fn parse_hex(i: &str) -> IResult<&str, u16> {
    lexeme(parse_hex_)(i)
}

pub fn parse_dec(i: &str) -> IResult<&str, u16> {
    lexeme(parse_dec_)(i)
}

pub fn parse_number(i: &str) -> IResult<&str, u16> {
    lexeme(parse_number_)(i)
}

pub fn parse_comma_sep(i: &str) -> IResult<&str, ()> {
    lexeme(parse_comma_sep_)(i)
}

pub fn parse_range_sep(i: &str) -> IResult<&str, ()> {
    lexeme(parse_range_sep_)(i)
}

pub fn parse_label(i: &str) -> IResult<&str, &str> {
    lexeme(parse_label_)(i)
}

pub fn parse_symbol_reference(i: &str) -> IResult<&str, RefToken> {
    lexeme(parse_symbol_reference_)(i)
}
