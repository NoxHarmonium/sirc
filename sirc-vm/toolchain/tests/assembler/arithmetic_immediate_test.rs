use insta::assert_snapshot;
use nom_supreme::{
    error::ErrorTree,
    final_parser::{final_parser, Location},
};
use pretty_hex::{config_hex, HexConfig};
use toolchain::data::object::build_object;
use toolchain::parsers::shared::parse_tokens;
use toolchain::types::shared::Token;

static PARSER_INPUT: &str = r"
; Two argument - destination and source register the same
ADDI r1, #0
ADCI r2, #0x0
SUBI r3, #1
SBCI r4, #0x1
ANDI r5, #254
ORRI r6, #0xF0
XORI r7, #0xFA
CMPI ll, #0xF
TSAI ah, #100
TSXI al, #0x10
COPI sh, #0xF0
ADDI sl, #0
ADCI ph, #0x0
SUBI pl, #1

; Two argument - immediate source shift
ADDI r1, #0, LSL #0
ADCI r2, #0x0, LSR #1
SUBI r3, #1, ASL #2
SBCI r4, #0x1, ASR #3
ANDI r5, #254, RTL #4
ORRI r6, #0xF0, RTR #5
XORI r7, #0xFA, NUL #6
CMPI ll, #0xF, LSR #8
TSAI ah, #100, ASL #9
TSXI al, #0x10, ASR #10
COPI sh, #0xF0, RTL #11
ADDI sl, #0, RTR #12
ADCI ph, #0x0, NUL #13
SUBI pl, #1, NUL #14
ADDI r1, #0, LSL #15

; Two argument - register source shift
ADDI r1, #0, LSL r1
ADCI r2, #0x0, LSR r2
SUBI r3, #1, ASL r3
SBCI r4, #0x1, ASR r4
ANDI r5, #254, RTL r5
ORRI r6, #0xAF, RTR r6
XORI r7, #0xFA, NUL r7
";

#[test]
fn test_assembler_arithmetic_immediate() {
    let hex_config = HexConfig {
        width: 4,
        ..HexConfig::default()
    };

    let tokens = match final_parser::<&str, Vec<Token>, ErrorTree<&str>, ErrorTree<Location>>(
        parse_tokens,
    )(PARSER_INPUT)
    {
        Ok(tokens) => tokens,
        Err(error) => panic!("Error parsing file:\n{error}"),
    };

    let object = build_object(tokens, "UNIT_TEST".to_string(), PARSER_INPUT.to_string());
    assert_snapshot!(config_hex(&object.program.as_slice(), hex_config));
}
