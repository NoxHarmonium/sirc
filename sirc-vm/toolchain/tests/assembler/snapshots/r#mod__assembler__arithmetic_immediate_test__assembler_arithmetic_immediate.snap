---
source: toolchain/tests/assembler/arithmetic_immediate_test.rs
expression: "config_hex(&object.program.as_slice(), hex_config)"
---
Length: 144 (0x90) bytes
0000:   00 40 00 10   .@..
0004:   04 80 00 10   ....
0008:   08 c0 00 50   ...P
000c:   0d 00 00 50   ...P
0010:   11 40 3f 90   .@?.
0014:   15 80 3c 10   ..<.
0018:   19 c0 3e 90   ..>.
001c:   2a 40 03 d0   *@..
0020:   32 80 19 10   2...
0024:   3a c0 04 10   :...
0028:   3f 00 3c 10   ?.<.
002c:   03 40 00 10   .@..
0030:   07 80 00 10   ....
0034:   0b c0 00 50   ...P
0038:   80 40 04 10   .@..
003c:   84 80 08 50   ...P
0040:   88 c0 4c 90   ..L.
0044:   8d 00 50 d0   ..P.
0048:   91 7f 95 10   ....
004c:   95 bc 19 50   ...P
0050:   99 fe 81 90   ....
0054:   aa 43 ca 10   .C..
0058:   b2 99 0e 50   ...P
005c:   ba c4 12 90   ....
0060:   bf 3c 16 d0   .<..
0064:   83 40 1b 10   .@..
0068:   87 80 03 50   ...P
006c:   8b c0 43 90   ..C.
0070:   80 40 07 d0   .@..
0074:   80 40 24 50   .@$P
0078:   84 80 28 90   ..(.
007c:   88 c0 6c d0   ..l.
0080:   8d 00 71 10   ..q.
0084:   91 7f b5 50   ...P
0088:   95 ab f9 90   ....
008c:   99 fe a1 d0   ....
